zerver/views/auth.py

```diff
@@ -80,6 +80,57 @@ def redirect_to_subdomain_login_url():
     redirect_url = login_url + '?subdomain=1'
     return HttpResponseRedirect(redirect_url)
 
+def login_or_ask_for_registration(request, email, user_profile, full_name='',
+                                  invalid_subdomain=False):
+    # type: (HttpRequest, Text, UserProfile, Text, Optional[bool]) -> HttpResponse
+    if invalid_subdomain:
+        # Show login page with an error message
+        return redirect_to_subdomain_login_url()
+
+    elif user_profile is None or user_profile.is_mirror_dummy:
+        # Ask for registration. If the user wants to register,
+        # send them to PreregistrationUser flow.
+        try:
+            validate_email(email)
+            invalid_email = False
+        except ValidationError:
+            # If email address is invalid, we can't send the user to
+            # PreregistrationUser flow.
+            invalid_email = True
+
+        url = reverse('zerver.views.auth.confirm_continue_registration')
+        context = {'full_name': full_name,
+                   'email': email,
+                   'current_url': lambda: url,
+                   'invalid_email': invalid_email,
+                   }
+        return render(request,
+                      'zerver/confirm_continue_registration.html',
+                      context=context)
+    else:
+        # This assertion makes sure that user_profile was authenticated
+        # through the authenticate method of Django which adds the
+        # backend property to user_profile.
+        assert hasattr(user_profile, 'backend')
+        login(request, user_profile)
+        if settings.REALMS_HAVE_SUBDOMAINS and user_profile.realm.subdomain is not None:
+            return HttpResponseRedirect(user_profile.realm.uri)
+        return HttpResponseRedirect("%s%s" % (settings.EXTERNAL_URI_SCHEME,
+                                              request.get_host()))
+
+@require_post
+@has_request_variables
+def confirm_continue_registration(request, full_name=REQ(), email=REQ()):
+    # type: (HttpRequest, Text, Text) -> HttpResponse
+    # The fact that we have reached here means that user_profile was None.
+    user_profile = None  # type: Optional[UserProfile]
+    return login_or_register_remote_user(
+        request,
+        email,
+        user_profile,
+        full_name=full_name,
+    )
```

Comment:
Right, it's not a CSRF issue.  The problem is that an attacker can go through the flow with their legitimate account until they reach the form that posts to here, and then replace their email address with a different one (e.g. in the DOM) before posting to this endpoint.  

The result would let them login or register using someone else's email.